File: packages/zyplot/android/src/main/java/com/hzblj/zyplot/charts/interaction/ChartSlots.kt
package com.hzblj.zyplot.charts.interaction

import androidx.compose.runtime.Composable
import androidx.compose.runtime.key
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.layout.layout
import androidx.compose.ui.unit.IntSize
import com.hzblj.zyplot.charts.plotRect
import com.hzblj.zyplot.core.ChartConfiguration
import kotlin.math.roundToInt

/**
 * Puts every node the app supplied where the chart says it belongs.
 *
 * The anchor is read inside the placement block rather than in composition, so a finger moving
 * costs a placement and not a recomposition — the same trade the tooltip the chart draws makes.
 */
@Composable
internal fun ChartSlots(
  config: ChartConfiguration,
  scrub: ChartScrub,
  width: Float,
  height: Float,
  density: Float,
  slot: @Composable (id: String, modifier: Modifier) -> Unit,
) {
  val plot = plotRect(config, width, height, density)
  /**
   * Kept past the touch so a node that fades out fades where it was read rather than at the corner.
   * Whether there is anything to show once the finger has gone is the app's to answer — it renders
   * the node or it does not — so the chart only ever answers where. Deliberately not snapshot
   * state: it is written in the placement block, which must not invalidate what it is placing.
   */
  val lastReading = remember { LastReading() }

  for (spot in annotationSpots(config, plot)) {
    key(spot.id) {
      val align = config.annotationViewAlign[spot.id]
      slot(
        "annotation:${spot.id}",
        Modifier.anchored { size ->
          when (spot.run) {
            SpotRun.ACROSS -> Offset(plot.left, alignedY(spot.at.y, size, align))
            SpotRun.DOWN -> Offset(spot.at.x - size.width / 2f, alignedY(plot, size, align))
            SpotRun.POINT -> Offset(spot.at.x - size.width / 2f, alignedY(spot.at.y, size, align))
          }
        },
      )
    }
  }

  // The plot arrives in pixels, so the anchor's own measurements have to leave dp to meet the ones
  // iOS holds in points.
  val anchor = config.tooltipAnchor
  val gap = anchor.gap * density
  val lift = anchor.lift * density

  slot(
    "tooltip",
    Modifier.anchored { size ->
      val at = pointer(scrub, lastReading) ?: return@anchored null
      if (anchor.isAbove) chip(at, size, plot, width, lift) else reading(at, size, plot, gap, anchor.align)
    },
  )

  slot(
    "range",
    Modifier.anchored { size ->
      val span = scrub.range.value ?: return@anchored null
      val start = categoryCentre(config, span.startIndex, width, density, align = "start")
      val end = categoryCentre(config, span.endIndex, width, density, align = "end")
      chip(Offset((start + end) / 2f, 0f), size, plot, width, lift)
    },
  )
}

/**
 * Where a view sits against a mark that has no height of its own — a point, or a rule that runs
 * across the plot. Centred on it unless asked otherwise: "top" puts its foot on the mark so it sits
 * above, "bottom" puts its head there so it hangs below.
 */
private fun alignedY(y: Float, size: IntSize, align: String?): Float = when (align) {
  "top" -> y - size.height
  "bottom" -> y
  else -> y - size.height / 2f
}

/**
 * Where a view sits against a rule that runs down the plot. That rule is a mark with a height of its
 * own, so the three are its head, its middle and its foot — and its head is where the chart's own
 * badge goes, which is what a view for one gets unless it asks for another.
 */
private fun alignedY(plot: Rect, size: IntSize, align: String?): Float = when (align) {
  "center" -> plot.center.y - size.height / 2f
  "bottom" -> plot.bottom - size.height
  else -> plot.top
}

private fun pointer(scrub: ChartScrub, lastReading: LastReading): Offset? {
  if (scrub.range.value != null) return null
  return scrub.pointer.value?.also { lastReading.at = it } ?: lastReading.at
}

/**
 * Centred on the reading and lifted clear of the plot, where the rule draws its own chip. Pinned to
 * the view rather than to the plot: the headroom the chip sits in is outside the plot's own box, and
 * a chip read at either end still belongs over the chart.
 */
private fun chip(pointer: Offset, size: IntSize, plot: Rect, viewWidth: Float, lift: Float): Offset {
  val centred = pointer.x - size.width / 2f

  return Offset(
    x = centred.coerceIn(0f, maxOf(0f, viewWidth - size.width)),
    y = plot.top - size.height - lift,
  )
}

private class LastReading {
  var at: Offset? = null
}

/**
 * Beside the reading where there is room for it, and on its other side where there is not.
 *
 * Down the plot it sits where the anchor says: against the top by the gap, which is where a card read
 * as belonging to the reading goes, or halfway down the plot, or against its floor.
 */
private fun reading(pointer: Offset, size: IntSize, plot: Rect, gap: Float, align: String?): Offset {
  val trailing = pointer.x + gap
  val x = if (trailing + size.width <= plot.right) trailing else pointer.x - gap - size.width
  val y = when (align) {
    "center" -> plot.center.y - size.height / 2f
    "bottom" -> plot.bottom - size.height - gap
    else -> plot.top + gap
  }

  return Offset(
    x = x.coerceIn(plot.left, maxOf(plot.left, plot.right - size.width)),
    y = y,
  )
}

/**
 * Places the node at whatever the anchor says, given its own measured size, and leaves it
 * unplaced — and so undrawn — when the anchor says nothing. Contributes no size of its own, so a
 * node the chart moves never grows the plot it sits on.
 */
private fun Modifier.anchored(anchor: (size: IntSize) -> Offset?): Modifier = layout { measurable, constraints ->
  val placeable = measurable.measure(constraints.copy(minHeight = 0, minWidth = 0))
  layout(0, 0) {
    val at = anchor(IntSize(placeable.width, placeable.height)) ?: return@layout
    placeable.place(at.x.roundToInt(), at.y.roundToInt())
  }
}

File: packages/zyplot/src/shared/chart-slots.ts
import type {ChartTooltipAnchor, NativeChartInteraction} from '@hzblj/zyplot-core'
import type {ReactNode} from 'react'
import {type ChartAnnotationViews, type ChartSlotView, hasAnnotationView, slotNode} from './annotation-views'

/**
 * Your own node in place of the chart's own tooltip. The chart keeps its own out of the
 * drawing and puts yours where it would have gone, moving it with the finger itself rather
 * than reporting the position back to you.
 *
 * On iOS and Android the node is mounted inside the chart, so it follows the reading in the
 * same frame the crosshair does. What is inside it is still yours to render from
 * `useChartScrub`, and that part arrives when React gets to it.
 *
 * Every slot takes an element or a component. A component is what a `zyplot` config can hold — it
 * is the same value on every render, so the config is too, and a reading that only changes what the
 * view shows never reaches the chart's props.
 *
 * @example
 * const {onInteraction, selection} = useChartScrub()
 * return (
 *   <Chart.Candlestick
 *     data={candles}
 *     onInteraction={onInteraction}
 *     tooltip={tooltip.beside({view: QuoteCard})}
 *   />
 * )
 */
export type ChartSlotViewProps = {
  /**
   * Your own view where an annotation lands, keyed by its `id`. The chart measures the
   * annotation, keeps its own mark for it out of the drawing, and centres your node on
   * the spot instead — a logo at the live reading, a badge on a rule, a whole card.
   *
   * Everything else about the annotation still applies: it moves with the data, and the
   * ones you leave out keep the dot, glow and pulse the chart draws.
   *
   * @example
   * zyplot(z => ({
   *   annotations: [z.annotation.point({id: 'live', x: live.category, y: live.value})],
   *   annotationViews: {live: LivePrice},
   * }))
   */
  annotationViews?: ChartAnnotationViews
  /**
   * Your own node for the span under two fingers, centred on it and lifted clear of the plot. The
   * chart draws no words for a span of its own, so this one adds rather than replaces.
   *
   * Needs `interaction.range`, and iOS or Android to be the one drawing — a pointer has no second
   * finger, so the web never places it.
   */
  rangeView?: ChartSlotView
  /**
   * What appears for the reading, in one setting. `false` for nothing at all; `true` or left out for
   * the card the chart writes itself; a `tooltip.beside({view})` or `tooltip.above({view})` for one
   * of your own, which the chart mounts and moves in its own layout pass.
   *
   * There is no second switch: a view here is the chart's card not being drawn, and a view placed
   * `above` is the crosshair's own label not being written.
   */
  tooltip?: boolean | ChartTooltip
}

/** The reading's own view, and the box the chart measures to place it. */
export type ChartTooltip = ChartTooltipAnchor & {view?: ChartSlotView}

/**
 * The interaction as the renderers read it: the group the app wrote, with what the `tooltip` prop
 * decided folded back in. `tooltip` is not on the group the app writes — one name for the reading is
 * the whole point — but it is what both renderers and the bridge have always been told.
 */
export type ResolvedChartInteraction = NativeChartInteraction & {tooltip?: boolean}

/** Where a node the app supplied is put, and what it is. */
export type ChartSlot = {
  id: string
  node: ReactNode
}

/** The slot the chart's own reading UI would have filled. */
export const TOOLTIP_SLOT = 'tooltip'

/** The slot for the span under two fingers. */
export const RANGE_SLOT = 'range'

const ANNOTATION_PREFIX = 'annotation:'

/** The slot an annotation's own mark would have filled. */
export const annotationSlot = (id: string): string => `${ANNOTATION_PREFIX}${id}`

/**
 * Every node the app supplied, in the order the chart mounts them. Entries written as
 * `live && <Now />` are left out, so a slot whose node has gone keeps whatever the chart
 * draws for it rather than losing both.
 */
const viewOf = (tooltip: boolean | ChartTooltip | undefined): ChartSlotView =>
  typeof tooltip === 'object' ? tooltip.view : undefined

/**
 * Where the chart is to put the app's own view for the reading, if it was given one.
 *
 * Named field by field rather than spread, because the `view` beside them cannot cross to a renderer.
 * Every field the anchor grows has to be added here too, or it is written and never arrives.
 */
export const tooltipAnchorOf = (tooltip: boolean | ChartTooltip | undefined): ChartTooltipAnchor | undefined =>
  typeof tooltip === 'object'
    ? {align: tooltip.align, gap: tooltip.gap, lift: tooltip.lift, placement: tooltip.placement}
    : undefined

export const chartSlots = ({annotationViews, rangeView, tooltip}: ChartSlotViewProps): readonly ChartSlot[] => {
  const slots: ChartSlot[] = []

  if (annotationViews) {
    for (const id of Object.keys(annotationViews).sort()) {
      const view = annotationViews[id]
      if (hasAnnotationView(view)) {
        slots.push({id: annotationSlot(id), node: slotNode(view)})
      }
    }
  }

  if (hasAnnotationView(viewOf(tooltip))) {
    slots.push({id: TOOLTIP_SLOT, node: slotNode(viewOf(tooltip))})
  }

  if (hasAnnotationView(rangeView)) {
    slots.push({id: RANGE_SLOT, node: slotNode(rangeView)})
  }

  return slots
}

/**
 * The interaction the renderers are handed: what the app asked for, plus what the `tooltip` prop
 * settles. The chart draws its own card unless it was told otherwise, and whatever the app's own view
 * stands in for is taken out — a view placed `'above'` is the rule's chip, so the chip's labels go.
 * Everything else the app asked for — the rule, the marker, the dimming — is untouched.
 */
export const resolveInteraction = (
  interaction: NativeChartInteraction | undefined,
  tooltip: boolean | ChartTooltip | undefined
): ResolvedChartInteraction | undefined => {
  const hasView = hasAnnotationView(viewOf(tooltip))

  if (tooltip === undefined && !hasView) {
    return interaction
  }

  const resolved: ResolvedChartInteraction = {...interaction, tooltip: tooltip !== false && !hasView}

  return hasView && typeof tooltip === 'object' && tooltip.placement === 'above' && resolved.crosshairStyle?.labels
    ? {...resolved, crosshairStyle: {...resolved.crosshairStyle, labels: undefined}}
    : resolved
}
